src/Language/PureScript/Sugar/Names.hs

```diff
@@ -244,21 +244,24 @@ renameInModule imports (Module modSS coms mn decls exps) =
     return (b, d)
 
   updateValue
-    :: (SourceSpan, [Ident])
+    :: (SourceSpan, M.Map Ident SourcePos)
     -> Expr
-    -> m ((SourceSpan, [Ident]), Expr)
+    -> m ((SourceSpan, M.Map Ident SourcePos), Expr)
   updateValue (_, bound) v@(PositionedValue pos' _ _) =
     return ((pos', bound), v)
   updateValue (pos, bound) (Abs (VarBinder ss arg) val') =
-    return ((pos, arg : bound), Abs (VarBinder ss arg) val')
+    return ((pos, M.insert arg (spanStart ss) bound), Abs (VarBinder ss arg) val')
   updateValue (pos, bound) (Let w ds val') = do
     let args = mapMaybe letBoundVariable ds
     unless (length (ordNub args) == length args) .
       throwError . errorMessage' pos $ OverlappingNamesInLet
-    return ((pos, args ++ bound), Let w ds val')
-  updateValue (_, bound) (Var ss name'@(Qualified Nothing ident)) | ident `notElem` bound =
-    ((ss, bound), ) <$> (Var ss <$> updateValueName name' ss)
-  updateValue (_, bound) (Var ss name'@(Qualified (Just _) _)) =
+    return ((pos, declarationsToMap ds `M.union` bound), Let w ds val')
+  updateValue (_, bound) (Var ss name'@(Qualified (BySourcePos _) ident)) =
+    ((ss, bound), ) <$> case M.lookup ident bound of
+      Just sourcePos -> pure $ Var ss (Qualified (BySourcePos sourcePos) ident)
+      Nothing -> Var ss <$> updateValueName name' ss
```

Comment:
Does this happen? I'd expect this branch to be an internal error.

Reply:
I believe it does, since the `ident` here is qualified by a null source position by default. I can do a quick trace to check, though.

Reply:
Oh, I think we miscommunicated. I meant the `Nothing` branch on this one line, which should only happen if `ident` is not in `bound` and is also not qualified by a module; I think those two conditions should be exhaustive. I think the `ByNullSourcePos` was better as a `BySourcePos _` pattern, though if a non-null `SourcePos` gets in there, maybe we should error on that too.

Reply:
Gotcha